contracts/integrations/adapters/ZeroExV3Adapter.sol

```diff
@@ -2,153 +2,158 @@
-    /// - [2] Protocol Fee asset (WETH)
-    /// - [3] Taker Fee asset (same as _orderAddresses[7])
-    /// @return fillExpectedAmounts_ Asset fill amounts
-    /// - [0] Expected (min) quantity of maker asset to receive
-    /// - [1] Expected (max) quantity of taker asset to spend
-    /// - [2] Expected (max) quantity of protocol asset to spend
-    /// - [3] Expected (max) quantity of taker fee asset to spend
-    /// @return fillApprovalTargets_ Recipients of assets in fill order
-    /// - [0] Taker (fund), set to address(0)
-    /// - [1] 0x asset proxy for the taker asset
-    /// - [2] 0x protocolFeeCollector
-    /// - [3] 0x asset proxy for the taker fee asset
-    function __formatFillTakeOrderArgs(bytes memory _encodedArgs)
-        internal
-        view
-        override
-        returns (address[] memory, uint256[] memory, address[] memory)
-    {
-        (
-            ,uint256[7] memory orderValues,
-            bytes[4] memory orderData,
-        ) = __decodeTakeOrderArgs(_encodedArgs);
-
-        address[] memory fillAssets = new address[](4);
-        fillAssets[0] = __getAssetAddress(orderData[0]); // maker asset
-        fillAssets[1] = __getAssetAddress(orderData[1]); // taker asset
-        fillAssets[2] = __getNativeAssetAddress(); // protocol fee
-        fillAssets[3] = __getAssetAddress(orderData[3]); // taker fee asset
-
-        uint256[] memory fillExpectedAmounts = new uint256[](4);
-        fillExpectedAmounts[0] = __calculateRelativeQuantity(
-            orderValues[1],
-            orderValues[0],
-            orderValues[6]
-        ); // maker fill amount; calculated relative to taker fill amount
-        fillExpectedAmounts[1] = orderValues[6]; // taker fill amount
-        fillExpectedAmounts[2] = __calcProtocolFeeAmount(); // protocol fee
-        fillExpectedAmounts[3] = __calculateRelativeQuantity(
-            orderValues[1],
-            orderValues[3],
-            orderValues[6]
-        ); // taker fee amount; calculated relative to taker fill amount
 
-        address[] memory fillApprovalTargets = new address[](4);
-        fillApprovalTargets[0] = address(0); // Fund (Use 0x0)
-        fillApprovalTargets[1] = __getAssetProxy(orderData[1]); // 0x asset proxy for taker asset
-        fillApprovalTargets[2] = IZeroExV3(EXCHANGE).protocolFeeCollector(); // 0x protocol fee collector
-        fillApprovalTargets[3] = __getAssetProxy(orderData[3]); // 0x asset proxy for taker fee asset
+        // Approve spend assets
+        __increaseApproval(
+            __getAssetAddress(order.takerAssetData),
+            __getAssetProxy(order.takerAssetData),
+            takerAssetFillAmount
+        );
 
-        return (fillAssets, fillExpectedAmounts, fillApprovalTargets);
-    }
+        uint256 protocolFee = __calcProtocolFeeAmount();
+        if (protocolFee > 0) {
+            __increaseApproval(
+                Registry(REGISTRY).nativeAsset(),
```

Comment:
Could be `IRegistry`

Reply:
Could be, which would make the deployment less expensive, but for now the convention is for all plugins to use full contracts for Melon core contracts. The idea is that the narrow interfaces used by our core contracts should reveal the surface area of interactions within the core system. I'm open to re-thinking the way we use interfaces, and whether plugins should use them, but for now let's stick to the convention.